### ISSUE TITLE
Toggle Keybinding

### ISSUE THREAD

**Issue by heph3astus** [dexterleng/homerow#36]
**Describe the solution you'd like**
Having a separate command to keep the actions and search on the screen after you've executed a click. This is great for going through menus or using this as a complete mouse replacement.

**Describe alternatives you've considered**
Currently just pressing the cmd+shft+space shortcut repeatedly

**Comment by dexterleng** [comment]
@heph3astus I would love that option also, one approach is just to auto reactivate Homerow after a click with a delay ~100ms. I don't know how well it would work in practice, I think I'll just expose some AppleScript events and implement this from the "outside" just to experiment without doing major refactors on my end for something that may not pan out.

Some questions I have:
1. Have you considered using an "easier" shortcut? e.g. `Caps Lock + F`, `Caps Lock + Space`, just `Caps Lock`, or maybe even Double tap Command? It would be great to have these shortcuts built into Homerow, but for now you could use Karabiner: https://github.com/dexterleng/homerow#pro-tip-map-caps-lock-to-hyper-key
2. Do you use the Tutor via mouse feature? If Homerow re-activated itself in cases where you don't want it to, I would imagine you would not want the Tutor to start showing a popover for whatever your cursor hovers over.

**Comment by heph3astus** [comment]
I do use the mouse occasionally for tutor, but rarely. Personally I would want anything to appear while I have the toggle engaged, although I think mouse disabling might be a nice feature. My goal is complete mouse replacement in certain apps. Also, I use a custom keyboard without a dedicated caps lock key so this is about the best I can do, but its also pretty easy to engage. Really its just the time sink of pressing buttons repeatedly. I might try setting something up to press the keybinding, then wait for enter, and press it again.

**Comment by inthegreenwild** [comment]
This would be a great option to toggle on or off. I usually want to keep homerow on while browsing

**Comment by goldlil** [comment]
@dexterleng you can send distrubuted notification when homerow perform major actions eg elementClicked or labelsCanceled and than with hammerspoon we can listen to those notifications and implement this. I have actually made this for vimac but its kinda bugy bcs my scrip is waiting for two letters to be entered after vimac shorcut was activated and after 1500ms it will acticate vemac again. It lable has more or less letters this scrip fails. But with distributed notifications this can works.

**Comment by dexterleng** [comment]
@goldlil Yes, that is the plan :) I'll share an early build here once I'm done.

**Comment by dexterleng** [comment]
Mind giving this debug build a try? It sends distributed notifications which you can listen for to create custom workflows.

There are two notifications:
1. `com.dexterleng.Homerow.debug.didActivate`
2. `com.dexterleng.Homerow.debug.didDeactivate`, which contains a `wasClickPerformed` property.3. 

[Download homerow-distributed-notifications.zip](https://github.com/dexterleng/homerow/files/10608847/homerow-distributed-notifications.zip)

I've written a [Hammerspoon script to perform auto-reactivation](https://gist.github.com/dexterleng/9b94613a69145c8672a7ec04d1ad2ff6).

I see two issues right now with auto-reactivation:
1. Websites can take a longer time to load than the delay (0.5s)
2. You don't want Homerow to reactivate itself when a menu is open, since activating Homerow will close the menu.

**Comment by dexterleng** [comment]
cc @nriley

**Comment by nriley** [comment]
Thanks!

Since I have mostly moved over to using search (for which you don't need any tags visible to start using Homerow), and I have [Rango](https://github.com/david-tejada/rango) for browser use I have not been wishing for this feature as much as I used to.

Talon does not have an interface to macOS distributed notifications. I've been [watching the focused element via accessibility](https://github.com/nriley/knausj_talon/blob/1fc1b4c3505b8aee8498a392cd08e7d279cc0442/apps/homerow/homerow.py#L51) to determine when Homerow deactivates, but sometimes this does not trigger immediately.

I do use Hammerspoon as well so I guess I could potentially bridge from Hammerspoon to Talon, but that seems like a lot :-).

**Comment by dexterleng** [comment]
@nriley

> Talon does not have an interface to macOS distributed notifications. I've been watching the focused element via accessibility to determine when Homerow deactivates, but sometimes this does not trigger immediately.

I see, other than accessibility, how do other Talon scripts listen for events from apps? Similarly, how do they trigger actions in apps? I tried implementing URL scheme in Homerow (`homerow://toggle`) but it causes the main app to lose focus which I don't want. Sending hotkeys is fine in Hammerspoon/Talon isn't the worse, but would be cool if there was a better approach.

> Talon does not have an interface to macOS distributed notifications. I've been watching the focused element via accessibility to determine when Homerow deactivates, but sometimes this does not trigger immediately.

I just gave your script a try - I don't fully understand the code. Why do you listen for `win_open` and `element_focus` events?

**Comment by nriley** [comment]
> I see, other than accessibility, how do other Talon scripts listen for events from apps? Similarly, how do they trigger actions in apps? I tried implementing URL scheme in Homerow (homerow://toggle) but opening a URL scheme causes the main app to lose focus which I don't want. Sending hotkeys is fine in Hammerspoon/Talon isn't the worse, but would be cool if there was a better approach.

There is a "missing RPC system" that will come at some point, but in the meantime, semi-structured RPC between Talon and apps occurs in practice:

1) over the clipboard + a trigger — e.g., an obscure keyboard shortcut Talon listens for to tell it to read the clipboard contents
2) through a file on disk — Talon has a file watcher that will invoke a callback if a file changes

The second option probably makes the most sense since it's pretty trivial to implement, if you wanted to expose Homerow's status in a temporary file somewhere.

On the Mac, Talon can send Apple Events using py-appscript. So if you could implement "show/hide/toggle Homerow search" and "search for this in Homerow" scripting commands, it'd drastically simplify/improve the reliability of my interaction with Homerow.

> I just gave your script a try - I don't fully understand the code. Why do you listen for win_open and element_focus events?
The speech command I am using to click on something ("pick …") doesn't normally do that outside Homerow; instead, it only does so when a particular tag is set. For example in the video below:

1) I say "ax assets", which searches Homerow for "assets"; there is only one result, so I just need to use "pick". I have the scope debug window open so you can see that `user.homerow_search` appears while the Homerow search window is open.

2) I say "ax icon", which searches for "icon"; there are multiple results. When I say "pick each", that causes Talon to press capital "E" and Return. 

3) I say "ax", which just pops up Homerow.  If I close the Homerow window manually ("escape") rather than with "pick", then `user.homerow_search` persists a little longer because there's been no `element_focus` event. This is not a huge deal but it may in a pathological case trigger a misrecognition with "pick". However, if Homerow could notify Talon it's no longer active, I wouldn't have this problem.

https://user-images.githubusercontent.com/68303/216838371-6e64630e-527d-4dcc-9b24-e5185b46bb53.mov

**Comment by meg-tong** [comment]
Hi, just wanted to say that I really like Homerow and would be excited about this feature. It would be great to have a keyboard shortcut to toggle between Homerow always-on and always-off.

I currently have it mapped to Shift-Space which is comparatively quick, but I'm usually doing a few things sequentially so it starts to feel slow.

**Comment by meg-tong** [comment]
FYI I tried the debug build and set up Hammerspoon with your script, but it'll show
```
2023-02-15 10:19:34: Homerow was activated.
2023-02-15 10:19:36: Homerow was deactivated. wasClickPerformed=	true
2023-02-15 10:19:36: Reactivating Homerow.
```
in the console but then I don't see Homerow actually reactivating. 
I haven't had time to look into it much; happy to try things if you have suggestions.

**Comment by GetBitOutdoors** [comment]
I would like this also, for Talon usage. Rango already does this. ShortCat tries to do this also, but doesnt seem to interact with Talon as reliably.

**Comment by goldlil** [comment]
@dexterleng can you include this notification in the Alpha version?

**Comment by KennethWussmann** [comment]
Hey,
I would still vote for a very simple solution for the start, because I think it would elevate the browsing experience.
Just giving the user the option to reactivate Homerow after a configurable timeout would already be beneficial for most use-cases.
